### PR TITLE
feat(contracts): ValueRouter

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

- goal: use a unified `transferRemote` interface to send msg value from A to B agnostic or the hook/ism needed to do so
- design doc: https://www.notion.so/hyperlanexyz/Native-bridge-value-transfer-API-1126d35200d680a0aa42d653d335a446

### Drive-by changes

- overrideMsgValue and overrideGasLimit which actually override the respective fields

### Backward compatibility

Yes

### Testing

Unit

### CODE REVIEW

File: solidity/contracts/token/HypValue.sol [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    /**
+     * @inheritdoc TokenRouter
+     * @dev This contract doesn't hold value
+     */
+    function balanceOf(
+        address /* _account */
+    ) external pure override returns (uint256) {
+        return 0;
+    }
```
Comment (yorhodes) [inline solidity/contracts/token/HypValue.sol:146]: the `HypNative` treats this differently, another reason to dedupe
```
function balanceOf(
        address _account
    ) external view override returns (uint256) {
        return _account.balance;
    }
```

File: solidity/contracts/token/HypValue.sol [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        address _recipient,
+        uint256 _amount,
+        bytes calldata // no token metadata
+    ) internal virtual override {
+        Address.sendValue(payable(_recipient), _amount);
+    }
+
+    /**
+     * @inheritdoc TokenRouter
+     * @dev This contract doesn't hold value
```
Comment (yorhodes) [inline solidity/contracts/token/HypValue.sol:140]: collateral balanceOf doesnt return what it holds, it returns what the user holds

File: solidity/contracts/token/HypValue.sol [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    ) internal view returns (uint256) {
+        uint256 quote = _GasRouter_quoteDispatch(
+            _destination,
+            new bytes(0),
+            _hook
+        );
+        if (msg.value < _amount + quote) {
+            revert InsufficientValue(_amount + quote, msg.value);
+        }
+        return quote;
```
Comment (yorhodes) [inline solidity/contracts/token/HypValue.sol:162]: its odd to me that you return quote here and then also do `amount + quote` outside of this function too

File: solidity/test/token/HypValue.t.sol
Comment (yorhodes) [inline solidity/test/token/HypValue.t.sol:1]: its nice that you tested this in conjunction with the OP hook but these are not really unit tests

File: solidity/test/token/HypValue.t.sol [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        );
+
+        bytes memory messengerCalldata = abi.encodeCall(
+            ICrossDomainMessenger.relayMessage,
+            (
+                0,
+                address(valueHook),
+                address(ism),
+                _msgValue,
+                uint256(100_000),
```
Comment (yorhodes) [inline solidity/test/token/HypValue.t.sol:181]: pull out this amount into a constant

File: solidity/test/token/HypValue.t.sol [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                encodedHookData
+            )
+        );
+        vm.deal(address(portal), _msgValue);
+        IOptimismPortal.WithdrawalTransaction
+            memory withdrawal = IOptimismPortal.WithdrawalTransaction({
+                nonce: 0,
```
Comment (yorhodes) [inline solidity/test/token/HypValue.t.sol:188]: ditto

File: solidity/test/token/HypValue.t.sol [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        localValueRouter.enrollRemoteRouter(
+            DESTINATION,
+            address(remoteToken).addressToBytes32()
+        );
+        remoteValueRouter.enrollRemoteRouter(
+            ORIGIN,
+            address(localToken).addressToBytes32()
+        );
+
+        vm.deal(ALICE, 1000e18);
```
Comment (yorhodes) [inline solidity/test/token/HypValue.t.sol:79]: whats this amount?

File: solidity/contracts/token/HypValue.sol [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+// ============ External Imports ============
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+/**
+ * @title HypValue
+ * @author Abacus Works
+ * @notice This contract facilitates the transfer of value between chains using value transfer hooks
+ */
+contract HypValue is TokenRouter {
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/token/HypValue.sol:28]: why?

File: solidity/contracts/hooks/OPL2ToL1Hook.sol [hunk 8]
```diff
@@ -83,9 +87,12 @@ contract OPL2ToL1Hook is AbstractMessageIdAuthHook {
             (message.id(), metadata.msgValue(0))
         );
 
+        bytes memory metadataWithGasLimit = metadata.overrideGasLimit(
+            MIN_GAS_LIMIT
+        );
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/hooks/OPL2ToL1Hook.sol:92]: Given that we know what the destination ism is, this hook will have a better estimate for the gas overhead so we should override it in all cases.

File: solidity/contracts/hooks/libs/StandardHookMetadata.sol [hunk 9]
```diff
@@ -165,4 +165,42 @@ library StandardHookMetadata {
...
+     */
+    function overrideMsgValue(
+        bytes calldata _metadata,
+        uint256 _msgValue
+    ) internal view returns (bytes memory) {
+        return
+            formatMetadata(
+                _msgValue,
+                gasLimit(_metadata, 0),
+                refundAddress(_metadata, msg.sender),
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/hooks/libs/StandardHookMetadata.sol:183]: these are just default values in case the specific part of the metadata is null
true but memcopy is cheap

File: solidity/contracts/token/HypValue.sol [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+     * @dev use _hook with caution, make sure that this hook can handle msg.value transfer using the metadata.msgValue()
+     */
+    function transferRemote(
+        uint32 _destination,
+        bytes32 _recipient,
+        uint256 _amount,
+        bytes calldata _hookMetadata,
+        address _hook
+    ) external payable virtual override returns (bytes32 messageId) {
+        uint256 quote = _checkSufficientValue(_destination, _amount, _hook);
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/token/HypValue.sol:67]: I attempted that first but ended by overridding almost everything

File: solidity/contracts/token/HypValue.sol [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            _amount,
+            address(hook)
+        );
+        bytes memory hookMetadata = StandardHookMetadata.formatMetadata(
+            _amount,
+            destinationGas[_destination],
+            msg.sender,
+            ""
+        );
+
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/token/HypValue.sol:102]: Done, but needs a calldata being passed.

File: solidity/contracts/token/HypValue.sol (hunk 1 above)
Comment (aroralanuk) [reply solidity/contracts/token/HypValue.sol:146]: Disagreed, this will be create more confusion. You get value from a different source and now the value is not reflective of what you actually bridged over. I prefer to make it 0 for HypNative too.

File: solidity/contracts/token/HypValue.sol (hunk 3 above)
Comment (aroralanuk) [reply solidity/contracts/token/HypValue.sol:162]: fixed

File: solidity/contracts/token/HypValue.sol (hunk 1 above)
Comment (aroralanuk) [reply solidity/contracts/token/HypValue.sol:146]: Another thing, you go bridge to HypNative, and now your wallet will show double the value, one as native value and second as HypNative balance.

File: solidity/test/token/HypValue.t.sol
Comment (aroralanuk) [reply solidity/test/token/HypValue.t.sol:1]: Agreed

File: solidity/test/token/HypValue.t.sol (hunk 6 above)
Comment (aroralanuk) [reply solidity/test/token/HypValue.t.sol:79]: it's 10* TRANSFER_AMT, fixed it

File: solidity/contracts/hooks/libs/StandardHookMetadata.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/hooks/libs/StandardHookMetadata.sol:183]: 0 gas limit and msg.sender for refund address does not make sense to me

File: solidity/contracts/token/HypValue.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/token/HypValue.sol]: passing around calldata doesnt memcopy :)

File: solidity/contracts/token/HypValue.sol (hunk 1 above)
Comment (yorhodes) [reply solidity/contracts/token/HypValue.sol:152]: > Disagreed, this will be create more confusion. You get value from a different source and now the value is not reflective of what you actually bridged over. I prefer to make it 0 for HypNative too.

This is already supported in the UI for HypNative. I am adamant that we should be consistent. Making HypNative 0 would be a breaking change that would brick all existing deployed warp routes.

Comment (yorhodes) [reply solidity/contracts/token/HypValue.sol:152]: > Another thing, you go bridge to HypNative, and now your wallet will show double the value, one as native value and second as HypNative balance.

Your wallet does not group native balance and token balances together.

File: solidity/contracts/token/HypValue.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/token/HypValue.sol:28]: as a user I would not know from the name how `HypValue` and `HypNative` differ

[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/token/HypValue.sol]: done

File: solidity/contracts/hooks/libs/StandardHookMetadata.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/hooks/libs/StandardHookMetadata.sol:183]: what would you have in their place?

File: solidity/contracts/token/HypValue.sol (hunk 1 above)
Comment (aroralanuk) [reply solidity/contracts/token/HypValue.sol]: Disagreed, but don't care too much, changing to account.balance.

File: solidity/contracts/token/HypNative.sol [hunk 12]
```diff
@@ -17,90 +32,122 @@ contract HypNative is TokenRouter {
...
-        require(msg.value >= _amount, "Native: amount exceeds msg.value");
-        uint256 _hookPayment = msg.value - _amount;
-        return _transferRemote(_destination, _recipient, _amount, _hookPayment);
-    }
+    // ============ External Functions ============
 
     /**
      * @inheritdoc TokenRouter
-     * @dev uses (`msg.value` - `_amount`) as hook payment.
+     * @dev use _hook with caution, make sure that this hook can handle msg.value transfer using the metadata.msgValue()
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/token/HypNative.sol:65]: suggested by you here: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/4814#discussion_r1832993090

File: solidity/contracts/token/HypNativeCollateral.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/token/HypNativeCollateral.sol:1]: If HypNative remains HypNative, then what should I call the new router? I went for HypValue since it sounds the most intuitive, but you suggested renaming it to HypNative, so here we are.

File: solidity/test/token/HypNative.t.sol [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+        IOptimismPortal.WithdrawalTransaction
+            memory withdrawal = IOptimismPortal.WithdrawalTransaction({
+                nonce: MOCK_NONCE,
+                sender: L2_MESSENGER_ADDRESS,
+                target: address(l1Messenger),
+                value: _msgValue,
+                gasLimit: OP_BRIDGE_GAS_LIMIT,
+                data: messengerCalldata
+            });
+        portal.finalizeWithdrawalTransaction(withdrawal);
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/test/token/HypNative.t.sol]: fixed this

File: solidity/contracts/hooks/libs/StandardHookMetadata.sol [hunk 14]
```diff
@@ -165,4 +165,45 @@ library StandardHookMetadata {
...
+     * @notice Overrides the gas limit in the metadata if _gasLimit is higher than the current gas limit.
+     * @param _metadata encoded standard hook metadata.
+     * @param _gasLimit gas limit for the message.
+     * @return encoded standard hook metadata.
+     */
+    function overrideGasLimit(
+        bytes calldata _metadata,
+        uint256 _gasLimit
+    ) internal view returns (bytes memory) {
+        if (gasLimit(_metadata, 0) < _gasLimit) {
```
Comment (yorhodes) [inline solidity/contracts/hooks/libs/StandardHookMetadata.sol:198]: the name of this function would not imply to me that there is this semantic of only override if higher

File: solidity/test/isms/ExternalBridgeTest.sol [hunk 15]
```diff
@@ -115,7 +115,6 @@ abstract contract ExternalBridgeTest is Test {
     function test_postDispatch_revertWhen_insufficientValue() public {
         bytes memory encodedHookData = _encodeHookData(messageId, 0);
         originMailbox.updateLatestDispatchedId(messageId);
-        _expectOriginExternalBridgeCall(encodedHookData);
```
Comment (yorhodes) [inline solidity/test/isms/ExternalBridgeTest.sol:118]: what changed here?

File: solidity/contracts/token/HypNative.sol [hunk 16]
```diff
@@ -17,90 +32,120 @@ contract HypNative is TokenRouter {
...
-    ) external payable virtual override returns (bytes32 messageId) {
-        require(msg.value >= _amount, "Native: amount exceeds msg.value");
-        uint256 _hookPayment = msg.value - _amount;
+    ) public payable virtual override returns (bytes32 messageId) {
+        uint256 quote = _GasRouter_quoteDispatch(
+            _destination,
+            _hookMetadata,
+            _hook
+        );
+        require(msg.value >= _amount + quote, "HypNative: insufficient value");
```
Comment (yorhodes) [inline solidity/contracts/token/HypNative.sol:79]: why did you change the conditions here?
previously we deferred to the hook to enforce the quote and I think that is desirable

File: solidity/contracts/token/HypNative.sol [hunk 17]
```diff
@@ -17,90 +32,120 @@ contract HypNative is TokenRouter {
...
      * @inheritdoc TokenRouter
-     * @dev No-op because native amount is transferred in `msg.value`
-     * @dev Compiler will not include this in the bytecode.
+     * @dev No token metadata is needed for value transfers
      */
     function _transferFromSender(
         uint256
     ) internal pure override returns (bytes memory) {
-        return bytes(""); // no metadata
+        return bytes(""); // no token metadata
```
Comment (yorhodes) [inline solidity/contracts/token/HypNative.sol:127]: why did you change this?

File: solidity/contracts/token/HypNative.sol [hunk 18]
```diff
@@ -17,90 +32,120 @@ contract HypNative is TokenRouter {
...
     }
 
     /**
-     * @dev Sends `_amount` of native token to `_recipient` balance.
      * @inheritdoc TokenRouter
+     * @dev Sends the value to the recipient
```
Comment (yorhodes) [inline solidity/contracts/token/HypNative.sol:132]: why change this?

File: solidity/contracts/token/HypNative.sol [hunk 19]
```diff
@@ -17,90 +32,120 @@ contract HypNative is TokenRouter {
...
         Address.sendValue(payable(_recipient), _amount);
     }
 
+    /// @inheritdoc TokenRouter
+    function balanceOf(
+        address _account
+    ) external view override returns (uint256) {
+        return _account.balance;
+    }
+
```
Comment (yorhodes) [inline solidity/contracts/token/HypNative.sol:148]: why move this?

File: solidity/contracts/token/HypNativeCollateral.sol
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/token/HypNativeCollateral.sol:1]: you did not add a changeset

File: solidity/contracts/token/HypNativeCollateral.sol [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+import {TokenRouter} from "./libs/TokenRouter.sol";
+import {TokenMessage} from "./libs/TokenMessage.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+/**
+ * @title Hyperlane Native Token Router that extends ERC20 with remote transfer functionality.
+ * @author Abacus Works
+ * @dev Supply on each chain is not constant but the aggregate supply across all chains is.
+ */
+contract HypNativeCollateral is TokenRouter {
```
Comment (yorhodes) [inline solidity/contracts/token/HypNativeCollateral.sol:13]: I really think this should be able to share a majority of implementation with the `HypNative` contract

File: solidity/test/isms/ExternalBridgeTest.sol (hunk 15 above)
Comment (aroralanuk) [reply solidity/test/isms/ExternalBridgeTest.sol:118]: It throws an 'OutOfFunds` error before it gets a chance to call the externall bridge.

File: solidity/contracts/token/HypNative.sol (hunk 16 above)
Comment (aroralanuk) [reply solidity/contracts/token/HypNative.sol]: done

File: solidity/contracts/token/HypNative.sol (hunk 18 above)
Comment (aroralanuk) [reply solidity/contracts/token/HypNative.sol:131]: I didn't, this is a new addition. I think you're confusing between the HypNative and HypNativeCollateral.

File: solidity/contracts/hooks/libs/StandardHookMetadata.sol (hunk 14 above)
Comment (aroralanuk) [reply solidity/contracts/hooks/libs/StandardHookMetadata.sol:198]: removed

File: solidity/test/isms/ExternalBridgeTest.sol (hunk 15 above)
Comment (aroralanuk) [reply solidity/test/isms/ExternalBridgeTest.sol]: fixed with removal

**Review by yorhodes — APPROVED** [review]
